### ISSUE TITLE
Created users cannot perform admin tasks

### ISSUE THREAD

**Issue by jaspertandy** [verbb/social-login#18]
### Describe the bug

I have an issue where users created via this plugin, and promoted to have admin, are not able to perform some admin tasks (such as adding admin to other users). This is because users created by this plugin do not have a password, and so can't enter anything into the popup that comes with certain admin tasks.

I can't find anything in the documentation on how to circumvent this without using the root account (which isn't controlled by this plugin) to either manually set a password to users, or add that admin privilege.

What is the preferred way to go about this?

I have noticed that the login expiry popup has a password field _and_ an "or log in with social options" under it. Could it be that this is just missing from the admin popup? My thoughts were that creating a user via the social plugin should/could trigger the password reset email so that new users are encouraged to set a password, but it feels a little strange to force new users to create a password that they don't need.

### Steps to reproduce

1. Enable Force Activation general setting
2. Enable user registration checked
3. Create a working provider
4. Log out of root account and log in with the new provider
5. User account is created and you can log in straight away
6. Log out of the account
7. Log into the root user account and add Admin to the newly-created user
8. Log out of the root account
9. Log into the social-login account that you just added admin to
10. Go to users > root admin account, and try to copy the password reset URL
11. Observe password popup which is impossible to complete

### Craft CMS version

4.5.5

### Plugin version

1.0.9

### Multi-site?

no

**Comment by engram-design** [comment]
That’s a great point, and something I’d not considered. I’ll need to look at what our options are here on this front.

You’re right that when the session expires, that pop up shows the login buttons for Social Login again, and we might need to see if that can be done for the elevated session pop up. There’s also complications on that as well, as we don’t want _any_ user to elevate their session through simply logging into social media.

And having a password reset would be against the goals for the plugin to use another service as your login provider.

I’ll have to investigate, but for now that’s not going to work sorry.

**Comment by jaspertandy** [comment]
Thanks for coming back so quickly. Glad you agree, and I understand it's not a super easy one to solve. I can kinda circumvent it now by managing expectations that users created with this plugin can't be admins, but would be nice if they _could_ be. Will wait to hear back from you but if I can provide anything further to help with this issue, please tag me.

**Comment by engram-design** [comment]
Yeah not an amazing situation. Or at the very least if they want to be an admin, they have to create a password with Craft (which is just for the elevated session).

**Comment by jaspertandy** [comment]
Yeah but the problem is, if you're already logged in you need a password to change your password, as changing your own password prompts you for your current password. You would have to have the root admin/an existing passworded admin copy your password reset URL so that you could do it without elevated access. But then I think you'd also need to log out before visiting that link but I haven't tried that.

Also I just upgraded to Craft 4.7.1 and the problem persists there. I know we already knew it would but just because my current version is no longer 4.5.5

**Comment by Diewy** [comment]
Has there been any update or workaround for this issue?

Still running into the same limitation with social-login users not being able to perform certain admin tasks. Curious if there’s a preferred approach in the meantime.

**Comment by jaspertandy** [comment]
We just have some users set a password. Not exactly ideal, but not all our users need to do admin tasks so it's at least workable for us.

**Comment by engram-design** [comment]
So I recall looking into this, and the elevated prompt isn't something we can hook into and circumvent its need for a password on the user account. I think this is one of those cases where if you want a socially-connected admin for Craft, they're going to need a password as well.

While we could _maybe_ come up with our own solution to hack the elevated handling Craft does, I don't exactly feel good about that, and poses security risks to this plugin and its users, where people could gain elevated access through easy means.

**Comment by jaspertandy** [comment]
Hi @engram-design, I haven't tested this all the way through, but the elevated session modal calls UsersController->actionLoginModal, which ends up rendering _special/login.twig: https://github.com/craftcms/cms/blob/5.x/src/templates/_special/login.twig#L144 which does seem to have something you might be able to hook into so that might have changed.

Then looks like you could just poke the elevated session duration once you've verified them? Not sure how that works in practice but at least seems technically possible? https://github.com/craftcms/cms/blob/5.x/src/web/User.php#L383

**Comment by engram-design** [comment]
Now that's a great point - that'll be new in Craft 5 where they added support for 2FA, and other login handling. I'll take a look! I know that the session-expired modal was something we could (and do) hook into but the session-elevation modal isn't a full login pane, it's just a password prompt.

**Comment by martinhj** [comment]
> Now that's a great point - that'll be new in Craft 5 where they added support for 2FA, and other login handling. I'll take a look! I know that the session-expired modal was something we could (and do) hook into but the session-elevation modal isn't a full login pane, it's just a password prompt.

What is the status for this now with Craft 5 in mind? Is the session-elevation modal still a password prompt that cannot easily be hooked into? I have no idea how this work with the 2FA in Craft 5 - I assume this is referring to the 2FA only provided with Craft enterprise?

**Comment by engram-design** [comment]
@martinhj No further update on this I'm afraid, other than we won't be able to make use of the 2FA in Craft 5, as that's a separate thing.

We're still working on the concept of login providers for Social Login, which would act in a similar way of Craft's 2FA or Passkey passwordless login.

**Comment by DieterVanhove** [comment]
Hi! I just wanted to add some context because I’m running into a very similar situation.

In my case, Craft has 2FA required for all users. When a new user signs in for the first time via SSO (using Social Login), the account is created without a Craft password, which makes sense for social accounts. However, as soon as Craft requires the user to complete 2FA setup, the flow also expects the user to have a password. This leads to the same problem described here: the user can’t proceed because they never set (and don’t need) a Craft password.

For now, we’ve worked around this by automatically sending a password-reset email to every new SSO-created user so they can set a password right away. That allows them to continue through Craft’s 2FA setup process normally.

Just sharing our scenario in case it’s helpful.

**Comment by engram-design** [comment]
Appreciate the extra scenario, didn't actually realise you needed a password to set up 2FA, so that's useful to know.

**Comment by petehjfd** [comment]
Hi @engram-design - can I ask if there has been any progress on this?  We are looking at rolling out Control Panel login/registration with Microsoft Entra across our 40+ Craft sites.  I'm doing some testing on one site locally at the moment.  When I try to copy a user impersonation URL (an action that requires an elevated session), I can see there is a "Sign in with" option for Entra in the "Confirm your identity" modal, but clicking it simply reloads the page and I don't get the impersonation URL modal.

Thanks

**Comment by engram-design** [comment]
This thread is somewhat meandering now, but to follow up on the original points:

You could use an event to handle this yourself:

```php
use yii\base\Event;
use verbb\sociallogin\events\UserEvent;
use verbb\sociallogin\services\Users;

Event::on(Users::class, Users::EVENT_BEFORE_REGISTER, function(UserEvent $event) {
    $event->user->admin = true;
});
```

Of course, adding your own logic in there as appriate to elevate your users to an admin. I've also added a new setting for some providers - "Allow Admin Registration", which you can opt into. To get this early, run `composer require verbb/social-login:"dev-craft-5 as 2.0.12"`.

The whole 2FA is for another discussion.

As for the elevated modal login, that's tricky to deal with. There's a tiny bug which I think should fix the refresh page (in the above dev branch).

**Comment by engram-design** [comment]
Updated in [2.0.13](https://github.com/verbb/social-login/releases/tag/2.0.13).
